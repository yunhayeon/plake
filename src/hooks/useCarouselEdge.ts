import { useEffect, useState } from "react";
import { useSwiper } from "swiper/react";

interface IUseCarouselEdgeProps {
  direction: "prev" | "next";
}

const useCarouselEdge = ({ direction }: IUseCarouselEdgeProps) => {
  const swiper = useSwiper();
  const [isEdge, setIsEdge] = useState(direction === "prev" ? true : false);

  useEffect(() => {
    const updateEdgeStatus = () => {
      setIsEdge(direction === "prev" ? swiper.isBeginning : swiper.isEnd);
    };

    updateEdgeStatus();

    swiper.on("slideChange", updateEdgeStatus);
    swiper.on("snapGridLengthChange", updateEdgeStatus);

    return () => {
      swiper.off("slideChange", updateEdgeStatus);
      swiper.off("snapGridLengthChange", updateEdgeStatus);
    };
  }, [swiper, direction]);

  return { isEdge, swiper };
};

export default useCarouselEdge;
