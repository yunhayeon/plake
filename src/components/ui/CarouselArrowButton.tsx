import clsx from "clsx";
import { IoIosArrowBack, IoIosArrowForward } from "react-icons/io";

import useCarouselEdge from "@/hooks/useCarouselEdge";

const CarouselArrowButton = ({ direction }: { direction: "prev" | "next" }) => {
  const { isEdge, swiper } = useCarouselEdge({ direction });

  return (
    <button
      onClick={() =>
        direction === "prev" ? swiper.slidePrev() : swiper.slideNext()
      }
      className={clsx(
        "",
        direction === "prev" ? "left-2" : "right-2",
        isEdge && "hidden",
        "group absolute top-1/2 z-10 -translate-y-1/2 rounded-full bg-white/80 p-2 shadow-md transition-all duration-200 hover:bg-white",
      )}
    >
      {direction === "prev" ? (
        <IoIosArrowBack className="-translate-x-[2px] text-2xl text-gray-800 group-hover:text-gray-600" />
      ) : (
        <IoIosArrowForward className="translate-x-[2px] text-2xl text-gray-800 group-hover:text-gray-600" />
      )}
    </button>
  );
};

export default CarouselArrowButton;
