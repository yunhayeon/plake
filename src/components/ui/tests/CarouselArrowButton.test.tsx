import { fireEvent, render, screen } from "@testing-library/react";

import useCarouselEdge from "@/hooks/useCarouselEdge";

import CarouselArrowButton from "../CarouselArrowButton";

jest.mock("@/hooks/useCarouselEdge", () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe("CarouselArrowButton 테스트", () => {
  let mockSwiper: {
    slidePrev: jest.Mock;
    slideNext: jest.Mock;
  };

  beforeEach(() => {
    jest.clearAllMocks();

    mockSwiper = {
      slidePrev: jest.fn(),
      slideNext: jest.fn(),
    };
  });

  describe("렌더링 테스트", () => {
    it("prev 버튼이 올바르게 렌더링", () => {
      (useCarouselEdge as jest.Mock).mockReturnValue({
        isEdge: false,
        swiper: mockSwiper,
      });

      render(<CarouselArrowButton direction="prev" />);
      const button = screen.getByRole("button");
      const icon = screen.getByLabelText("prev-arrow");

      expect(button).toMatchSnapshot();
      expect(icon).toBeInTheDocument();
    });

    it("next 버튼이 올바르게 렌더링", () => {
      (useCarouselEdge as jest.Mock).mockReturnValue({
        isEdge: false,
        swiper: mockSwiper,
      });

      render(<CarouselArrowButton direction="next" />);
      const button = screen.getByRole("button");
      const icon = screen.getByLabelText("next-arrow");

      expect(button).toMatchSnapshot();
      expect(icon).toBeInTheDocument();
    });

    it("isEdge가 true일 때 버튼이 숨겨짐", () => {
      (useCarouselEdge as jest.Mock).mockReturnValue({
        isEdge: true,
        swiper: mockSwiper,
      });

      render(<CarouselArrowButton direction="prev" />);
      const button = screen.queryByRole("button");

      expect(button).toHaveClass("hidden");
    });
  });

  describe("이벤트 테스트", () => {
    it("prev 버튼 클릭 시 slidePrev 호출", () => {
      (useCarouselEdge as jest.Mock).mockReturnValue({
        isEdge: false,
        swiper: mockSwiper,
      });

      render(<CarouselArrowButton direction="prev" />);
      const button = screen.getByRole("button");

      fireEvent.click(button);
      expect(mockSwiper.slidePrev).toHaveBeenCalled();
    });

    it("next 버튼 클릭 시 slideNext 호출", () => {
      (useCarouselEdge as jest.Mock).mockReturnValue({
        isEdge: false,
        swiper: mockSwiper,
      });

      render(<CarouselArrowButton direction="next" />);
      const button = screen.getByRole("button");

      fireEvent.click(button);
      expect(mockSwiper.slideNext).toHaveBeenCalled();
    });
  });
});
