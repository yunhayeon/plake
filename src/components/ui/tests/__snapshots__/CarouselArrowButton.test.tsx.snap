// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselArrowButton 테스트 렌더링 테스트 next 버튼이 올바르게 렌더링 1`] = `
<button
  class="right-2 group absolute top-1/2 z-10 -translate-y-1/2 rounded-full bg-white/80 p-2 shadow-md transition-all duration-200 hover:bg-white"
>
  <svg
    aria-label="next-arrow"
    class="translate-x-[2px] text-2xl text-gray-800 group-hover:text-gray-600"
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 512 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M294.1 256L167 129c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.3 34 0L345 239c9.1 9.1 9.3 23.7.7 33.1L201.1 417c-4.7 4.7-10.9 7-17 7s-12.3-2.3-17-7c-9.4-9.4-9.4-24.6 0-33.9l127-127.1z"
    />
  </svg>
</button>
`;

exports[`CarouselArrowButton 테스트 렌더링 테스트 prev 버튼이 올바르게 렌더링 1`] = `
<button
  class="left-2 group absolute top-1/2 z-10 -translate-y-1/2 rounded-full bg-white/80 p-2 shadow-md transition-all duration-200 hover:bg-white"
>
  <svg
    aria-label="prev-arrow"
    class="-translate-x-[2px] text-2xl text-gray-800 group-hover:text-gray-600"
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 512 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M217.9 256L345 129c9.4-9.4 9.4-24.6 0-33.9-9.4-9.4-24.6-9.3-34 0L167 239c-9.1 9.1-9.3 23.7-.7 33.1L310.9 417c4.7 4.7 10.9 7 17 7s12.3-2.3 17-7c9.4-9.4 9.4-24.6 0-33.9L217.9 256z"
    />
  </svg>
</button>
`;
